### Brief Overview 

- The goal of this assignment is to practice some important data wrangling functionality commonly required in real-world projects.

- Here we will use two datasets:
  - IRS Statistics of Income (SOI) dataset
  - The Medicaid Data per State  


- The final product here is a table with medication cost per Medicaid enrollee per state. This dataset will allow us to answer such questions as:
  - Which medications account for the bulk of a state's spending   
  - Which drugs are prescribed much more in one state compared to the other states.
etc.



### Instructions on answering the questions

- Some of the questions below require that you only use methods or properties of a `DataFrame` or a `Series`. Therefore, any solution that uses a function that is not a method or property of `DataFrame`s or `Series` will not be accepted, even when the solution yields an appropriate answer to the question.

- For instance, if you are asked to find the number of entries in the DataFrame `tax_data ` using only the DataFrame's methods or properties, then `len(tax_data)` is not an acceptable solution since `len()` is not a `tax_data` method. The statements below are both correct answers:

  - `info()` is a method `tax_data`
  
```python
   
    tax_data.info()
```

  or

  - `shape` is a property of `tax_data`


```python
tax_data.shape
```

- Similarly, if you are asked to find to count the number of unique entries in the `STATEFIPS` column of the `tax_data` DataFrame, then solutions using set() or len() are not acceptable for the following reasons:

- The solution below uses `set()` and `len()`, which are not `tax_data` methods

```python
len(set(tax_data["STATEFIPS"]))
```

- The solution below uses `unique()`, which is a  `tax_data` method, but then counts the number of uniques entries using `len()` which is not `tax_data` method

```python
len(tax_data["STATEFIPS"].unique())
```


- The statement below is an acceptable solution since it uses `nunique()` which is a method of the `Series` generated by indexing on a column of `tax_data` (`tax_data['STATEFIPS'])

```python
tax_data['STATEFIPS'].nunique()
```


- Chaining methods and properties is encouraged if it does not cause ambiguity 

- For instance, to identify whether a value is part of the index, write the following:

```python
tax_data.index.contains(99999)
```

- Rather than:

```python
9999 in tax_data.index
```

- You can only import `pandas` and `numpy`

- If you are not explicitely asked to only use methods or properties of a `DataFrame` or a `Series`, then any solution that does not rely on external products will be accepted.

- __Important__: Provide the exact statement(s) used to answer each question

- Unless otherwise specified, each cell can contain multiple lines of code

* Finally, note that not all the functions necessary for answering the questions below were covered in class. As such, I suggest you use `.SHIFT+TAB` on objects liberally to see which methods and properties are available on objects. If you are unsure what a method does, use `SHIFT+TAB` twice to invoke the `docstring`, or documentation for that function. This is not only a good way to see which functionality can be used to answer the questions below but also a great way to familiarize yourself with the plethora or functionality available through the `pandas` package.

---

In [1]:
import pandas as pd
import numpy as np


* Load the IRS Statistics of Income (SOI) dataset (tax_data.csv) into a `DataFrame` called `tax_data`. The file is `tax_data.csv` is located in the `data` directory of the assignment folder.

* This dataset was preprocessed but the original one was obtained at the following URL:

        https://www.irs.gov/pub/irs-soi/15zpallagi.csv


In [2]:
## WRITE YOUR CODE HERE 
tax_data = pd.read_csv('/Users/aditijaiswal/Documents/GitHub/data-wrangling-jaiswal-aditi/data/tax_data.csv')

* Use a `tax_data` method or property to display the first eight (8) rows of the `DataFrame`

In [3]:
## WRITE YOUR CODE HERE 
tax_data.head(8)

STATEFIPS STATE  zipcode  agi_stub        N1     mars1     MARS2     MARS4  \
0        1.0    AL      0.0       1.0  836320.0  481570.0  109790.0  233260.0   
1        1.0    AL      0.0       2.0  494830.0  206630.0  146250.0  129390.0   
2        1.0    AL      0.0       3.0  261250.0   80720.0  139280.0   36130.0   
3        1.0    AL      0.0       4.0  166690.0   28510.0  124650.0   10630.0   
4        1.0    AL      0.0       5.0  212660.0   19520.0  184320.0    4830.0   
5        1.0    AL      0.0       6.0   55360.0    2950.0   49260.0     350.0   
6        1.0    AL  35004.0       1.0    1490.0     970.0     230.0     280.0   
7        1.0    AL  35004.0       2.0    1350.0     630.0     360.0     300.0   

       PREP         N2  ...    N10300     A10300   N85530   A85530   N85300  \
0  455560.0  1356760.0  ...  373410.0   328469.0      0.0      0.0      0.0   
1  275920.0  1010990.0  ...  395880.0   965011.0      0.0      0.0      0.0   
2  155100.0   583910.0  ...  251490.0  1333418.0      0.0      0.0      0.0   
3   99950.0   423990.0  ...  165320.0  1414283.0      0.0      0.0      0.0   
4  126860.0   589490.0  ...  212000.0  3820152.0    420.0    168.0     60.0   
5   41410.0   160530.0  ...   55300.0  6027793.0  22090.0  39519.0  27550.0   
6     700.0     2160.0  ...     690.0      610.0      0.0      0.0      0.0   
7     610.0     2540.0  ...    1140.0     3019.0      0.0      0.0      0.0   

    A85300   N11901    A11901    N11902     A11902  
0      0.0  61920.0   48150.0  732670.0  1933120.0  
1      0.0  73720.0  107304.0  415410.0  1187403.0  
2      0.0  64200.0  139598.0  193030.0   536699.0  
3      0.0  45460.0  128823.0  116440.0   377177.0  
4     31.0  83330.0  421004.0  121570.0   483682.0  
5  95112.0  28590.0  791573.0   15960.0   250289.0  
6      0.0    120.0      94.0    1290.0     2792.0  
7      0.0    210.0     301.0    1130.0     2935.0  

[8 rows x 91 columns]

*  Modify `tax_data` to uppercase all the header name. 
  * Your answer can only use `DataFrame` or `Series` methods or properties
  * Do not hardcode the operation by uppercasing the columns yourself
  *  You can `tax_data.columns`, which returns a `Series` of the column names.
  
* The resulting column name should look as follows:


```
STATEFIPS    STATE    ZIPCODE    AGI_STUB    N1    MARS1    MARS2    MARS4    PREP    N2    ...    
```

This operation is useful for standardizing column names and avoid guessing whether the column header was in upper case, lower case or a mix of both.

In [4]:
## WRITE YOUR CODE HERE 
tax_data.columns = tax_data.columns.str.upper()
tax_data.head()

STATEFIPS STATE  ZIPCODE  AGI_STUB        N1     MARS1     MARS2     MARS4  \
0        1.0    AL      0.0       1.0  836320.0  481570.0  109790.0  233260.0   
1        1.0    AL      0.0       2.0  494830.0  206630.0  146250.0  129390.0   
2        1.0    AL      0.0       3.0  261250.0   80720.0  139280.0   36130.0   
3        1.0    AL      0.0       4.0  166690.0   28510.0  124650.0   10630.0   
4        1.0    AL      0.0       5.0  212660.0   19520.0  184320.0    4830.0   

       PREP         N2  ...    N10300     A10300  N85530  A85530  N85300  \
0  455560.0  1356760.0  ...  373410.0   328469.0     0.0     0.0     0.0   
1  275920.0  1010990.0  ...  395880.0   965011.0     0.0     0.0     0.0   
2  155100.0   583910.0  ...  251490.0  1333418.0     0.0     0.0     0.0   
3   99950.0   423990.0  ...  165320.0  1414283.0     0.0     0.0     0.0   
4  126860.0   589490.0  ...  212000.0  3820152.0   420.0   168.0    60.0   

   A85300   N11901    A11901    N11902     A11902  
0     0.0  61920.0   48150.0  732670.0  1933120.0  
1     0.0  73720.0  107304.0  415410.0  1187403.0  
2     0.0  64200.0  139598.0  193030.0   536699.0  
3     0.0  45460.0  128823.0  116440.0   377177.0  
4    31.0  83330.0  421004.0  121570.0   483682.0  

[5 rows x 91 columns]

* What is the total number of entries (also called observations) in `tax_data`?

  * Your answer can only use `DataFrame` or `Series` methods or properties


In [5]:
## WRITE YOUR CODE HERE 
tax_data.shape[0]

166698

- If `STATEFIPS` is header title of the first column of the `tax_data` `DataFrame`, what is the title of the 32nd column
  - Your answer can only use `DataFrame` or `Series` methods or properties and should use a single python expression



In [6]:
## WRITE YOUR CODE HERE 
tax_data.columns[31]

'N00900'

* If `STATEFIPS` is the the first column, what is the index of the column name `N10300`?

  * Your answer can only use `DataFrame` or `Series` methods or properties


In [7]:
## WRITE YOUR CODE HERE 
tax_data.columns.get_loc('N10300')

81

- Print the count of unique zip codes in each state using descending order.
  - Your answer can only use `DataFrame` or `Series` methods or properties

- The result should look like the following ( '...' represents remaining data that is not shown)
```python

        ZIPCODE
STATE	
TX	     9714
NY	     9238
CA	     8908
PA	     8208
IL	     7386
...
```

* The above indicates that there are 9,714 unique zipcodes in TX, 9,238 in NY, etc.


In [8]:
## WRITE YOUR CODE HERE 
unique_zipcode = tax_data.groupby(['STATE'])['ZIPCODE'].nunique().sort_values(ascending=False)
unique_zipcode

STATE
TX    1619
NY    1540
CA    1485
PA    1368
IL    1231
OH     998
FL     920
MI     892
MO     890
IA     827
VA     793
MN     792
NC     725
WI     714
IN     674
GA     667
KY     656
KS     598
TN     590
AL     577
NJ     547
OK     547
WV     512
WA     497
AR     490
NE     486
MA     482
LA     453
MD     405
CO     395
SC     376
MS     370
ME     370
OR     356
AZ     295
ND     287
SD     287
CT     264
VT     238
NH     232
MT     225
ID     213
NM     209
UT     186
NV     128
WY     109
RI      71
HI      60
DE      57
AK      56
DC      24
Name: ZIPCODE, dtype: int64

- Identify the position of HI in the list of zip code counts per state (questions directly above)
  - Your answer can only use `DataFrame` or `Series` methods or properties

In [9]:
## WRITE YOUR CODE HERE 
unique_zipcode.index.get_loc('HI')

47

### Identifying and Removing Ambiguous Zip Codes

- Count the number of entries where ZIPCODE is 0, assign your results to a variable named  `nb_invalid_zip`

In [10]:
## WRITE YOUR CODE HERE 
nb_invalid_zip = tax_data[tax_data['ZIPCODE']==0].shape[0]
nb_invalid_zip
#type(nb_invalid_zip)

306

* Run the line below to make sure that `nb_invalid_zip` is an integer (`int`)
  * Note that `assert` will only print an error if `type(nb_invalid_zip)` is not of type `int`
  
* If `nb_invalid_zip` then change your answer above so that the value returned is effectively a number.

In [11]:
assert(type(nb_invalid_zip) == int)

- Remove from `tax_data` all the lines where the zip code is `0` and save resulting `DataFrame` to a variable named `tax_data_valid_zip`
  - Your answer can only use `DataFrame` or `Series` methods or properties


In [12]:
## WRITE YOUR CODE HERE 
tax_data_valid_zip = tax_data[tax_data['ZIPCODE'] != 0]
tax_data_valid_zip.head(10)

STATEFIPS STATE  ZIPCODE  AGI_STUB      N1  MARS1  MARS2  MARS4   PREP  \
6         1.0    AL  35004.0       1.0  1490.0  970.0  230.0  280.0  700.0   
7         1.0    AL  35004.0       2.0  1350.0  630.0  360.0  300.0  610.0   
8         1.0    AL  35004.0       3.0   970.0  310.0  490.0  140.0  450.0   
9         1.0    AL  35004.0       4.0   620.0  110.0  470.0   30.0  300.0   
10        1.0    AL  35004.0       5.0   620.0   40.0  560.0   20.0  320.0   
11        1.0    AL  35004.0       6.0    60.0    0.0   50.0    0.0   50.0   
12        1.0    AL  35005.0       1.0  1350.0  750.0  190.0  390.0  730.0   
13        1.0    AL  35005.0       2.0   980.0  370.0  230.0  350.0  530.0   
14        1.0    AL  35005.0       3.0   490.0  150.0  210.0  120.0  290.0   
15        1.0    AL  35005.0       4.0   250.0   50.0  180.0   40.0  150.0   

        N2  ...  N10300   A10300  N85530  A85530  N85300  A85300  N11901  \
6   2160.0  ...   690.0    610.0     0.0     0.0     0.0     0.0   120.0   
7   2540.0  ...  1140.0   3019.0     0.0     0.0     0.0     0.0   210.0   
8   2160.0  ...   930.0   5009.0     0.0     0.0     0.0     0.0   200.0   
9   1610.0  ...   620.0   5190.0     0.0     0.0     0.0     0.0   150.0   
10  1770.0  ...   620.0  10129.0     0.0     0.0     0.0     0.0   220.0   
11   150.0  ...    60.0   3680.0     0.0     0.0    20.0    30.0    30.0   
12  2300.0  ...   610.0    545.0     0.0     0.0     0.0     0.0   100.0   
13  1970.0  ...   770.0   1679.0     0.0     0.0     0.0     0.0   100.0   
14  1070.0  ...   460.0   2373.0     0.0     0.0     0.0     0.0   100.0   
15   620.0  ...   250.0   2098.0     0.0     0.0     0.0     0.0    50.0   

    A11901  N11902  A11902  
6     94.0  1290.0  2792.0  
7    301.0  1130.0  2935.0  
8    376.0   760.0  2058.0  
9    374.0   460.0  1390.0  
10   854.0   410.0  1709.0  
11   357.0     0.0     0.0  
12    74.0  1200.0  3243.0  
13   116.0   880.0  2501.0  
14   226.0   390.0  1136.0  
15   133.0   200.0   650.0  

[10 rows x 91 columns]

* Run the line below to confirm that the operation worked as expected

  * The assertion below is testing that the number of lines with "zip code equal to  0" + number of lines in `tax_data_valid_zip` is equal to the number of lines in the original `DataFrame` `tax_data`
  
  * The assertion below will fail (and print an error message) if the results do not match. If that is the case, please review your code above.

In [13]:
assert((tax_data_valid_zip.shape[0] + nb_invalid_zip) == tax_data.shape[0])

### Identifying and Removing Lines with Missing Values

* How many lines contain at least one missing value `NaN` in the `tax_data_valid_zip` `DataFrame`?
  * Your answer can only use `DataFrame` methods and properties
* Assing the count of `NaN` into a variable called nb_missing_values

In [14]:
## WRITE YOUR CODE HERE 
nb_missing_values = tax_data_valid_zip.isna().any(axis=1).sum()
nb_missing_values

139

* Create a new `DataFrame` containing all the lines from `tax_data_valid_zip`, except lines containing missing values

* Call the new `DataFrame` `tax_data_valid_zip_cleaned

In [15]:
## WRITE YOUR CODE HERE 
tax_data_valid_zip_cleaned = tax_data_valid_zip.dropna(axis=0)

* Run the line below to confirm that the operation worked as expected. The assertion below is testing that:  
`nb_missing_values` + number of lines in `tax_data_valid_zip_cleaned` is equal to the number of lines in `tax_data_valid_zip`
  
* Note that assert will only print an error if the results do not match

In [16]:
assert((tax_data_valid_zip_cleaned.shape[0] + nb_missing_values) == tax_data_valid_zip.shape[0])

### Computing the Percentile Income per Zip Code

* The function `compute_percentile_zipcode` below computes the percentile income per zip code

* By default percentile=0.5,  i.e., the function computes the median

* Read the code and make sure you understand what it does before moving on to the next question


In [17]:
def compute_percentile_zip(df_zip, percentile=0.5):     
    index_median = sum(( df_zip["N1"]/ sum(df_zip["N1"])).cumsum() <= percentile)
    val_below_or_at_median = (df_zip["A00100"] /df_zip["N1"]).iloc[index_median]
    return val_below_or_at_median

* Compute the 65th income percentile ( percentile=0.65) for each zipcode in `tax_data_valid_zip_cleaned` data frame

  * Your answer can only use `DataFrame` or `Series` methods or properties
  
  * Recall that you can use the `apply` function to transform the values of a column
  
  
* Sort the values in descending order and assign them to a new `DataFrame` called `zip_rev_all`
* The resulting `Series` should look like the following( '...' represents remaining data that is not shown)

```Python
ZIPCODE
33109    3954.114286
33480    3413.301538
...
```

In [18]:
## WRITE YOUR CODE HERE 
zip_rev_all = tax_data_valid_zip_cleaned.groupby('ZIPCODE').apply(compute_percentile_zip, 0.65)
zip_rev_all = zip_rev_all.sort_values(ascending=False)
zip_rev_all

ZIPCODE
33109.0    3954.114286
33480.0    3413.301538
94301.0    3109.443711
94027.0    3091.537013
10577.0    2414.855556
              ...     
53706.0       7.292857
55455.0       7.210000
48109.0       6.981250
84112.0       6.597297
59053.0       5.421429
Length: 27682, dtype: float64

- What are the three zip codes with the most significant 65th percentile value for income?

In [19]:
## WRITE YOUR CODE HERE 
zip_rev_all.iloc[:3]

ZIPCODE
33109.0    3954.114286
33480.0    3413.301538
94301.0    3109.443711
dtype: float64

# 2 Working with the Medicaid Data

### Loading and exploring the data 

* Load the Medicaid data stored in the file `medicaid_data.csv` into a `DataFrame` called `medicaid_data`. The file is located in the `data` directory of the assignment folder. 
* Note that this is quite large and may take some time to load on a computer with modest RAM resources (4GB or less)


In [20]:
## WRITE YOUR CODE HERE 
medicaid_data = pd.read_csv('/Users/aditijaiswal/Downloads/medicaid_data.csv')
medicaid_data.head()

Utilization Type State          NDC Product Name  Units Reimbursed  \
0             MCOU    PA  55150023930   Dexamethas              33.0   
1             FFSU    NY     23917710   ALPHAGAN P             570.0   
2             MCOU    OR  13925050501   Dapsone 10             456.0   
3             FFSU    MN  51862006401     DIAZEPAM             780.0   
4             FFSU    MN    781237101   DEXTROAMPH             451.0   

   Number of Prescriptions  Total Amount Reimbursed  \
0                     19.0                   234.98   
1                     57.0                 16006.34   
2                     15.0                  1052.42   
3                     16.0                    89.60   
4                     12.0                  1411.24   

   Medicaid Amount Reimbursed  Non Medicaid Amount Reimbursed  \
0                      234.98                            0.00   
1                    16006.34                            0.00   
2                     1052.42                            0.00   
3                       77.60                           12.00   
4                      198.93                         1212.31   

               Location  
0    (40.5773, -77.264)  
1   (42.1497, -74.9384)  
2  (44.5672, -122.1269)  
3   (45.7326, -93.9196)  
4   (45.7326, -93.9196)


- Modify `medicaid_data` to uppercase all the column names 

  - If your solution uses an assignment, the righthand side of the assignment (rvalue) can only use `DataFrame` or `Series` methods or properties


In [21]:
## WRITE YOUR CODE HERE 
medicaid_data.columns = medicaid_data.columns.str.upper()
medicaid_data.head()

UTILIZATION TYPE STATE          NDC PRODUCT NAME  UNITS REIMBURSED  \
0             MCOU    PA  55150023930   Dexamethas              33.0   
1             FFSU    NY     23917710   ALPHAGAN P             570.0   
2             MCOU    OR  13925050501   Dapsone 10             456.0   
3             FFSU    MN  51862006401     DIAZEPAM             780.0   
4             FFSU    MN    781237101   DEXTROAMPH             451.0   

   NUMBER OF PRESCRIPTIONS  TOTAL AMOUNT REIMBURSED  \
0                     19.0                   234.98   
1                     57.0                 16006.34   
2                     15.0                  1052.42   
3                     16.0                    89.60   
4                     12.0                  1411.24   

   MEDICAID AMOUNT REIMBURSED  NON MEDICAID AMOUNT REIMBURSED  \
0                      234.98                            0.00   
1                    16006.34                            0.00   
2                     1052.42                            0.00   
3                       77.60                           12.00   
4                      198.93                         1212.31   

               LOCATION  
0    (40.5773, -77.264)  
1   (42.1497, -74.9384)  
2  (44.5672, -122.1269)  
3   (45.7326, -93.9196)  
4   (45.7326, -93.9196)


- Familiarize your self with the data
  - the `NDC` column stands for National Drug Code, a universal product identifier for human drugs in the United States
  
  - The remaining column names are self-explanatory
  
- Explore the number of lines and columns in the data

- Check that your column headers are in uppercase

In [22]:
## WRITE YOUR CODE HERE 
print(medicaid_data.shape)
medicaid_data.columns

(1695546, 10)


Index(['UTILIZATION TYPE', 'STATE', 'NDC', 'PRODUCT NAME', 'UNITS REIMBURSED',
       'NUMBER OF PRESCRIPTIONS', 'TOTAL AMOUNT REIMBURSED',
       'MEDICAID AMOUNT REIMBURSED', 'NON MEDICAID AMOUNT REIMBURSED',
       'LOCATION'],
      dtype='object')

* If you explore the  `Location` column for all the entries for which "STATE" value is equal to "HI" you'll notice that all the values are identical

* Are there any states that have more than one value for `Location`. 
  * Hint: think about using a sorted aggregation as part of a split-apply-combine operation to answer this question
  - Your answer can only use `DataFrame` or `Series` methods or properties


In [23]:
## WRITE YOUR CODE HERE 
#medicaid_data[medicaid_data["STATE"] == "HI"]["LOCATION"]
medicaid_data.groupby(['STATE', 'LOCATION']).size()

STATE  LOCATION            
AK     (61.385, -152.2683)     12265
AL     (32.799, -86.8073)      23547
AR     (34.9513, -92.3809)     18276
AZ     (33.7712, -111.3877)    27348
CA     (36.17, -119.7462)      76996
CO     (39.0646, -105.3272)    26067
CT     (41.5834, -72.7622)     24586
DC     (38.8964, -77.0262)     16099
DE     (39.3498, -75.5148)     13277
FL     (27.8333, -81.717)      49457
GA     (32.9866, -83.6487)     45797
HI     (21.1098, -157.5311)    13438
IA     (42.0046, -93.214)      22580
ID     (44.2394, -114.5103)    14626
IL     (40.3363, -89.0022)     52561
IN     (39.8647, -86.2604)     41411
KS     (38.5111, -96.8005)     19154
KY     (37.669, -84.6514)      45011
LA     (31.1801, -91.8749)     38406
MA     (42.2373, -71.5314)     43582
MD     (39.0724, -76.7902)     37519
ME     (44.6074, -69.3977)     14091
MI     (43.3504, -84.5603)     46286
MN     (45.7326, -93.9196)     44117
MO     (38.4623, -92.302)      30845
MS     (32.7673, -89.6812)     29863
MT     (46

In [24]:
medicaid_data.groupby('STATE')['LOCATION'].nunique()

STATE
AK    1
AL    1
AR    1
AZ    1
CA    1
CO    1
CT    1
DC    1
DE    1
FL    1
GA    1
HI    1
IA    1
ID    1
IL    1
IN    1
KS    1
KY    1
LA    1
MA    1
MD    1
ME    1
MI    1
MN    1
MO    1
MS    1
MT    1
NC    1
ND    1
NE    1
NH    1
NJ    1
NM    1
NV    1
NY    1
OH    1
OK    1
OR    1
PA    1
RI    1
SC    1
SD    1
TN    1
TX    1
UT    1
VA    1
VT    1
WA    1
WI    1
WV    1
WY    1
XX    0
Name: LOCATION, dtype: int64



* To compare medication prescriptions across states in a fair and balanced way, we need the number of Medicaid beneficiaries in each state. The following example illustrates the importance of normalizing the values `UNITS REIMBURSED` for each medication in each state by the number of Medicaid enrollees in each state.
  
* The `medicaid_data` DataFrame shows that for the drug with NDC `61958180101` (the drug name is HARVONI and it's used to treat Hepatitis C) there were 11,886  units sold in KY, versus 40,142 in CA -- that's almost 4 times more units sold in CA compared to KY. However, there are 1,284,193 Medicaid enrollees in KY, versus 13,096,861 in California. If we normalize the number of units sold in KY, versus CA, we find that the normalized there were close to 3 times more HARVONI prescription in KY  than in CA. This is ___perhaps___ justified by the fact the KY has one of the highest rates of reported cases of Hepatitis C in the US (2.7% in KY versus 0.2% in CA).
  
https://www.cdc.gov/hepatitis/statistics/2015surveillance/pdfs/2015hepsurveillancerpt.pdf

* The number of enrollees per state was obtained here:

    https://www.medicaid.gov/medicaid/managed-care/enrollment/index.html
    
    
* A parsed/processed version (medicaid_enrollment.tsv) can be in data director of the assignment folder. Use `pandas` to load the medicaid_enrollmen file into DataFrame named `medicaid_enrollment`

In [25]:
## WRITE YOUR CODE HERE 
medicaid_enrollment = pd.read_csv('/Users/aditijaiswal/Documents/GitHub/data-wrangling-jaiswal-aditi/data/medicaid_enrollment.tsv', sep='\t')
medicaid_enrollment.head()
# keep_default_na=False

STATE Total Medicaid Enrollees
0         Alabama                1,050,989
1          Alaska                  164,783
2  American Samoa                      NaN
3         Arizona                1,740,520
4        Arkansas                  762,166

* Modify `medicaid_enrollment` to uppercase all the column names 

  * Your answer can only use `DataFrame` or `Series` methods or properties
  * Do not hardcode the operation by uppercasing the columns yourself


In [26]:
## WRITE YOUR CODE HERE 
medicaid_enrollment.columns = medicaid_enrollment.columns.str.upper()
medicaid_enrollment.head()

STATE TOTAL MEDICAID ENROLLEES
0         Alabama                1,050,989
1          Alaska                  164,783
2  American Samoa                      NaN
3         Arizona                1,740,520
4        Arkansas                  762,166

* Note that some states/territories have missing values. Remove the missing values and save the resulting `DataFrame` as a new variable named `medicaid_enrollment_cleaned`

* Pay attention to how 'n/a' is given here!
* After cleaning, do you still have the Guam entry? If so, reconsider what missing values means in this context

In [27]:
## WRITE YOUR CODE HERE 
#medicaid_enrollment.isnull().sum()
#medicaid_enrollment_cleaned = medicaid_enrollment.dropna()
#medicaid_enrollment_cleaned = medicaid_enrollment_cleaned[medicaid_enrollment_cleaned["TOTAL MEDICAID ENROLLEES"].str.strip() != "n/a"]
#medicaid_enrollment_cleaned.head(20)

medicaid_enrollment_cleaned = medicaid_enrollment.replace({'n/a' : np.nan} ,regex=True).dropna()
medicaid_enrollment_cleaned.head(20)

STATE TOTAL MEDICAID ENROLLEES
0                Alabama                1,050,989
1                 Alaska                  164,783
3                Arizona                1,740,520
4               Arkansas                  762,166
5             California               13,096,861
6               Colorado                1,264,600
7            Connecticut                  746,119
8               Delaware                  227,909
9   District of Columbia                  271,428
10               Florida                3,808,334
11               Georgia                1,990,810
13                Hawaii                  340,513
14                 Idaho                  283,355
15              Illinois                3,269,999
16               Indiana                1,295,358
17                  Iowa                  618,505
18                Kansas                  403,844
19              Kentucky                1,284,193
20             Louisiana                1,402,212
21                 Maine                  288,324

### Converting `TOTAL MEDICAID ENROLLEE` Data Type

* Given that data on `TOTAL MEDICAID ENROLLEE` column contains commas on file (ex. 3,269,999 instead of 3269999), `pandas` has erroneously set the data type for that column as a string. We need to convert the column from string to `int` since we will be using it in an arithmetic expression during normalization

 

* Inspect the `dtype` property of "TOTAL MEDICAID ENROLLEES" column, and  make sure that the data type is `int`


In [28]:
## WRITE YOUR CODE HERE 
medicaid_enrollment_cleaned['TOTAL MEDICAID ENROLLEES'].dtypes

dtype('O')

In [29]:
medicaid_enrollment_cleaned['TOTAL MEDICAID ENROLLEES'] = medicaid_enrollment_cleaned['TOTAL MEDICAID ENROLLEES'].str.replace(',', '').astype(str).astype(int)
print(medicaid_enrollment_cleaned.head())
medicaid_enrollment_cleaned['TOTAL MEDICAID ENROLLEES'].dtypes

        STATE  TOTAL MEDICAID ENROLLEES
0     Alabama                   1050989
1      Alaska                    164783
3     Arizona                   1740520
4    Arkansas                    762166
5  California                  13096861


dtype('int64')

### Associating `medicaid_data` and `medicaid_enrollment_cleaned`

- We can use the shared State information across both tables to associate both tables (SQL JOIN).
- However,  `medicaid_data` contains two-letter state abbreviations, while `medicaid_enrollment_cleaned` contains the complete state name
  - We need to convert (or append) two-letter state abbreviations to `medicaid_enrollment_cleaned`

- Pandas can read HTML and parse the code for tables. We will use that functionality to read in the state abbreviations from a Wikipedia page.
  - A brief description of what the code does is included in the comments

In [30]:
'''# reads all the tables at the given url
# Header=0 instructs pandas to use line 0 as the header (columns)
tables = pd.read_html('https://www.50states.com/abbreviations.htm', header=0)


# We access the desired table by giving it's index.
# Since the URL contain only one table, then we can access that table using index 0
Codes_abbreviations = tables[0]
Codes_abbreviations.head(5)'''

import requests

url = 'https://www.50states.com/abbreviations.htm'
header = {
  "User-Agent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/50.0.2661.75 Safari/537.36",
  "X-Requested-With": "XMLHttpRequest"
}

r = requests.get(url, headers=header)

tables = pd.read_html(r.text)


# We access the desired table by giving it's index.
# Since the URL contain only one table, then we can access that table using index 0
Codes_abbreviations = tables[0]
Codes_abbreviations.head(5)

US STATE POSTAL ABBREVIATION STANDARD ABBREVIATION
0     Alabama                  AL                  Ala.
1      Alaska                  AK                Alaska
2     Arizona                  AZ                 Ariz.
3    Arkansas                  AR                  Ark.
4  California                  CA                Calif.

* Change the the `DataFrame`'s headers from ['US State:', 'Abbreviation:'] to ['US STATE', 'ABBREVIATION']

  * You can hard code this operation


In [31]:
## WRITE YOUR CODE HERE 
Codes_abbreviations.drop(['STANDARD ABBREVIATION'], axis=1, inplace=True)
Codes_abbreviations.columns = ["US STATE", 'ABBREVIATION']
Codes_abbreviations.head()

US STATE ABBREVIATION
0     Alabama           AL
1      Alaska           AK
2     Arizona           AZ
3    Arkansas           AR
4  California           CA

* Combine the tables `medicaid_enrollment_cleaned` and `Codes_abbreviations` such that the resulting `DataFrame` contains all the columns in `medicaid_enrollment_cleaned` and only `ABBREVIATION` from `Codes_abbreviations` 
* Save the results to variable named `medicaid_enrollment_cleaned_with_zip`

- `medicaid_enrollment_cleaned_with_zip` should look like the following ( '...' represents remaining data that is not shown):


```
   STATE    Total Medicaid Enrollees    ABBREVIATION
0    Alabama    1,050,989    AL
1    Alaska    164,783    AK
2    Arizona    1,740,520    AZ
3    Arkansas    762,166    AR
4    California    13,096,861    CA
...
```

* We did not cover joins in class -- you find a plethora of examples on how to do this online. See for instance:

`https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.merge.html`

* If you cannot get it to work, contact the `TA` for the solution. You will not be penalized if you don't answer this question. 

In [32]:
## WRITE YOUR CODE HERE 
medicaid_enrollment_cleaned_with_zip = medicaid_enrollment_cleaned.merge(Codes_abbreviations, how='left', left_on='STATE', right_on='US STATE')
medicaid_enrollment_cleaned_with_zip = medicaid_enrollment_cleaned_with_zip.drop('US STATE', axis=1)
medicaid_enrollment_cleaned_with_zip.head()

STATE  TOTAL MEDICAID ENROLLEES ABBREVIATION
0     Alabama                   1050989           AL
1      Alaska                    164783           AK
2     Arizona                   1740520           AZ
3    Arkansas                    762166           AR
4  California                  13096861           CA

* We have no further use for the column STATE in  `medicaid_enrollment_cleaned_with_zip`
  * Remove the column make sure your data in `medicaid_enrollment_cleaned_with_zip` looks like the following  ( `...` represents remaining data that is not shown):

```
    Total Medicaid Enrollees    ABBREVIATION
0   1,050,989                  AL
1   164,783                    AK
2   1,740,520                  AZ
3   762,166                    AR
4   13,096,861                 CA
....
```

In [33]:
## WRITE YOUR CODE HERE 
medicaid_enrollment_cleaned_with_zip = medicaid_enrollment_cleaned_with_zip.drop('STATE', axis=1)
medicaid_enrollment_cleaned_with_zip.head()

TOTAL MEDICAID ENROLLEES ABBREVIATION
0                   1050989           AL
1                    164783           AK
2                   1740520           AZ
3                    762166           AR
4                  13096861           CA

- Use `DataFrame medicaid_enrollment_cleaned_with_zip` to assign the appropriate number of Medicaid enrollees to each entry in the `medicaid_data`

   I.E., instead of the 10 original columns, `medicaid_data` will now have an 11th column representing the `TOTAL MEDICAID ENROLLEES` according to the STATE value in the entry.
   
- Save the resulting DataFrame into a new variable called `medicaid_data_w_enrollments`
- The resulting DataFrame should look like the following (`...` represents remaining data that is not shown):

```
UTILIZATION TYPE    STATE    NDC    PRODUCT NAME    UNITS REIMBURSED    NUMBER OF PRESCRIPTIONS    TOTAL AMOUNT REIMBURSED    MEDICAID AMOUNT REIMBURSED    NON MEDICAID AMOUNT REIMBURSED    LOCATION
0    MCOU    PA    55150023930    Dexamethas    33.0    19.0    234.98    234.98    0.0    (40.5773, -77.264)
1    FFSU    NY    23917710    ALPHAGAN P    570.0    57.0    16006.34    16006.34    0.0    (42.1497, -74.9384)
2    MCOU    OR    13925050501    Dapsone 10    456.0    15.0    1052.42    1052.42    0.0    (44.5672, -122.1269)
...
```

* The order of the columns in the `DataFrame` is not important. This answer uses the same approach as the one used to `merge` the tables above. 

In [34]:
## WRITE YOUR CODE HERE 
medicaid_data_w_enrollments = medicaid_data.merge(medicaid_enrollment_cleaned_with_zip, left_on='STATE', right_on='ABBREVIATION')
medicaid_data_w_enrollments.drop('ABBREVIATION', axis=1, inplace=True)
medicaid_data_w_enrollments.head()

UTILIZATION TYPE STATE          NDC PRODUCT NAME  UNITS REIMBURSED  \
0             MCOU    PA  55150023930   Dexamethas              33.0   
1             FFSU    PA    456060010   TEFLARO 60             224.0   
2             MCOU    PA    115692201   DIVALPROEX            1626.0   
3             FFSU    PA     54455025    METHOTREX             450.0   
4             MCOU    PA  50458014090     INVOKANA            3850.0   

   NUMBER OF PRESCRIPTIONS  TOTAL AMOUNT REIMBURSED  \
0                     19.0                   234.98   
1                     36.0                 35752.31   
2                     20.0                  2318.55   
3                     22.0                   565.29   
4                    111.0                 52655.13   

   MEDICAID AMOUNT REIMBURSED  NON MEDICAID AMOUNT REIMBURSED  \
0                      234.98                            0.00   
1                    35752.31                            0.00   
2                     2287.80                           30.75   
3                      516.23                           49.06   
4                    50641.22                         2013.91   

             LOCATION  TOTAL MEDICAID ENROLLEES  
0  (40.5773, -77.264)                   2569232  
1  (40.5773, -77.264)                   2569232  
2  (40.5773, -77.264)                   2569232  
3  (40.5773, -77.264)                   2569232  
4  (40.5773, -77.264)                   2569232

In [35]:
print(medicaid_data.shape)
medicaid_data_w_enrollments.shape

(1695546, 10)


(1526194, 11)

- Remove any lines where "STATE" or "PRODUCT NAME" are missing from  `medicaid_data_w_enrollments`

In [36]:
## WRITE YOUR CODE HERE 
medicaid_data_w_enrollments.isnull().any()

UTILIZATION TYPE                  False
STATE                             False
NDC                               False
PRODUCT NAME                       True
UNITS REIMBURSED                  False
NUMBER OF PRESCRIPTIONS           False
TOTAL AMOUNT REIMBURSED           False
MEDICAID AMOUNT REIMBURSED        False
NON MEDICAID AMOUNT REIMBURSED    False
LOCATION                          False
TOTAL MEDICAID ENROLLEES          False
dtype: bool

In [37]:
medicaid_data_w_enrollments = medicaid_data_w_enrollments.dropna(subset=['STATE', 'PRODUCT NAME'])
print(medicaid_data_w_enrollments.isnull().any())
medicaid_data_w_enrollments['PRODUCT NAME'].isnull().sum()

UTILIZATION TYPE                  False
STATE                             False
NDC                               False
PRODUCT NAME                      False
UNITS REIMBURSED                  False
NUMBER OF PRESCRIPTIONS           False
TOTAL AMOUNT REIMBURSED           False
MEDICAID AMOUNT REIMBURSED        False
NON MEDICAID AMOUNT REIMBURSED    False
LOCATION                          False
TOTAL MEDICAID ENROLLEES          False
dtype: bool


0

In [38]:
medicaid_data_w_enrollments.shape

(1526193, 11)

- Use ["STATE", "PRODUCT NAME"] as hierarchical index for `medicaid_data_w_enrollments`. Recall that a hierarchical index is simply an index with multiple levels of indexing (multiple columns)
  * Hint: the function to set an index on a `DataFrame` can take a single column name or a list of column names. The list here is  ["STATE", "NDC"]
- Call the new data `medicaid_data_w_enrollments_hierarch`
- Inspect your data to make sure the new index has now two levels (STATE and NDC)

In [39]:
## WRITE YOUR CODE HERE 
medicaid_data_w_enrollments_hierarch = medicaid_data_w_enrollments.set_index(["STATE", "NDC"])
medicaid_data_w_enrollments_hierarch

UTILIZATION TYPE PRODUCT NAME  UNITS REIMBURSED  \
STATE NDC                                                           
PA    55150023930             MCOU   Dexamethas             33.00   
      456060010               FFSU   TEFLARO 60            224.00   
      115692201               MCOU   DIVALPROEX           1626.00   
      54455025                FFSU    METHOTREX            450.00   
      50458014090             MCOU     INVOKANA           3850.00   
...                            ...          ...               ...   
NJ    63323031466             MCOU   Vancomycin             19.53   
      51991074810             MCOU   Duloxetine           3048.00   
      406051201               MCOU    OXYCODONE         161572.00   
      115681108               MCOU    BUPROPION           5098.00   
      50242004164             MCOU   CATHFLO AC            182.50   

                   NUMBER OF PRESCRIPTIONS  TOTAL AMOUNT REIMBURSED  \
STATE NDC                                                             
PA    55150023930                     19.0                   234.98   
      456060010                       36.0                 35752.31   
      115692201                       20.0                  2318.55   
      54455025                        22.0                   565.29   
      50458014090                    111.0                 52655.13   
...                                    ...                      ...   
NJ    63323031466                     14.0                   822.99   
      51991074810                     93.0                  4270.15   
      406051201                     3926.0                 34981.10   
      115681108                      160.0                  3810.15   
      50242004164                    155.0                 16620.63   

                   MEDICAID AMOUNT REIMBURSED  NON MEDICAID AMOUNT REIMBURSED  \
STATE NDC                                                                       
PA    55150023930                      234.98                            0.00   
      456060010                      35752.31                            0.00   
      115692201                       2287.80                           30.75   
      54455025                         516.23                           49.06   
      50458014090                    50641.22                         2013.91   
...                                       ...                             ...   
NJ    63323031466                      822.99                            0.00   
      51991074810                     4270.15                            0.00   
      406051201                      34791.07                          190.03   
      115681108                       3810.15                            0.00   
      50242004164                    15103.73                         1516.90   

                             LOCATION  TOTAL MEDICAID ENROLLEES  
STATE NDC                                                        
PA    55150023930  (40.5773, -77.264)                   2569232  
      456060010    (40.5773, -77.264)                   2569232  
      115692201    (40.5773, -77.264)                   2569232  
      54455025     (40.5773, -77.264)                   2569232  
      50458014090  (40.5773, -77.264)                   2569232  
...                               ...                       ...  
NJ    63323031466  (40.314, -74.5089)                   1705594  
      51991074810  (40.314, -74.5089)                   1705594  
      406051201    (40.314, -74.5089)                   1705594  
      115681108    (40.314, -74.5089)                   1705594  
      50242004164  (40.314, -74.5089)                   1705594  

[1526193 rows x 9 columns]


* Write a single Pandas expression to print all the lines with containing NDC 61958180101 in "PA"

 * Use a single indexing call (bracket notation) using `loc`
 
 * Hint 1: Since your index is hierarchical, `loc` is expecting two values, the first for STATE and the second for NDC


In [40]:
## WRITE YOUR CODE HERE 
medicaid_data_w_enrollments_hierarch.loc[('PA',  61958180101)]

/var/folders/65/zjdkrmxs3qnd_z9_8zt_vpr80000gn/T/ipykernel_6354/659043288.py:2: PerformanceWarning: indexing past lexsort depth may impact performance.
  medicaid_data_w_enrollments_hierarch.loc[('PA',  61958180101)]


UTILIZATION TYPE PRODUCT NAME  UNITS REIMBURSED  \
STATE NDC                                                           
PA    61958180101             FFSU   Harvoni  (             924.0   
      61958180101             MCOU   Harvoni  (            2604.0   
      61958180101             FFSU   Harvoni  (            1008.0   
      61958180101             MCOU   Harvoni  (            1932.0   
      61958180101             FFSU   Harvoni  (             924.0   
      61958180101             MCOU   Harvoni  (            3220.0   

                   NUMBER OF PRESCRIPTIONS  TOTAL AMOUNT REIMBURSED  \
STATE NDC                                                             
PA    61958180101                     33.0               1017644.54   
      61958180101                     93.0               2952891.16   
      61958180101                     36.0               1107725.22   
      61958180101                     69.0               2178602.30   
      61958180101                     33.0               1015383.60   
      61958180101                    115.0               3599192.35   

                   MEDICAID AMOUNT REIMBURSED  NON MEDICAID AMOUNT REIMBURSED  \
STATE NDC                                                                       
PA    61958180101                   985269.84                        32374.70   
      61958180101                  2860100.54                        92790.62   
      61958180101                  1107725.22                            0.00   
      61958180101                  2118612.30                        59990.00   
      61958180101                  1015383.60                            0.00   
      61958180101                  3482188.68                       117003.67   

                             LOCATION  TOTAL MEDICAID ENROLLEES  
STATE NDC                                                        
PA    61958180101  (40.5773, -77.264)                   2569232  
      61958180101  (40.5773, -77.264)                   2569232  
      61958180101  (40.5773, -77.264)                   2569232  
      61958180101  (40.5773, -77.264)                   2569232  
      61958180101  (40.5773, -77.264)                   2569232  
      61958180101  (40.5773, -77.264)                   2569232

#### Computing the normalized `UNITS REIMBURSED` per druc 

* Compute the `UNITS REIMBURSED` for each "NDC" in each state normalized by the number of enrollees.

- For instance in `PA`, the total `UNITS REIMBURSED` for the HARVONI `NDC` (61958180101) is 10,612

```python
total_amount_reimbursed = medicaid_data_w_enrollments_hierarch.loc[("PA", 61958180101), "UNITS REIMBURSED"].sum() 
```

- And the numebr of Medicaid Enrollees in "PA" is 2,569,232


```python
total_enrollees_PA = medicaid_data_w_enrollments_hierarch.loc["PA", "TOTAL MEDICAID ENROLLEES"].unique()[0]
```
- Therefore, the UNITS REIMBURSED per enrollee  for "HARVONI" is  0.00413041718303


```python
print(total_amount_reimbursed/total_enrollees_AK)
```

- Rather than work directly with the ratios, we are going to compute and work with thier logarithm (np.log2) instead.

  - The reason we use logs here is to avoid working small numbers. More on log-transformation in future lectures

- Save the result in a `Series` called `medicaid_reimbursement_per_enrollee`


- Your final result should be a Series that look like the following ( `...` represents data that is not shown here):
  
```
STATE  NDC    
AK     2143380    -9.609109
       2143480   -10.008280
       2322730    -6.109830
       2322830    -4.444321
       2322930    -3.855995
...

AL     2143380    -9.940595
       2143480    -9.805485
       2322730    -5.336260
...

MA     2143380    -7.921997
       2143480    -7.803463
       2144511   -13.472194
       2197590    -7.741402
       2322730    -5.414724
       2322830    -4.592154
       2322930    -4.205626

...


```


In [41]:
## WRITE YOUR CODE HERE 
units_medicaid = medicaid_data_w_enrollments_hierarch.groupby(level=[0,1])[['UNITS REIMBURSED']].sum()
#units_medicaid.head(20)
num_enrollee = medicaid_data_w_enrollments_hierarch['TOTAL MEDICAID ENROLLEES']
#enrollee.head(20)
temp = units_medicaid.join(num_enrollee, how='outer')
temp.head(10)

UNITS REIMBURSED  TOTAL MEDICAID ENROLLEES
STATE NDC                                                
AK    2143380             211.0                    164783
      2143380             211.0                    164783
      2143380             211.0                    164783
      2143480             160.0                    164783
      2143480             160.0                    164783
      2143480             160.0                    164783
      2322730            2386.0                    164783
      2322730            2386.0                    164783
      2322830            7569.0                    164783
      2322830            7569.0                    164783

In [42]:
clean_temp = temp.drop_duplicates()
clean_temp.head(20)

UNITS REIMBURSED  TOTAL MEDICAID ENROLLEES
STATE NDC                                                
AK    2143380           211.000                    164783
      2143480           160.000                    164783
      2322730          2386.000                    164783
      2322830          7569.000                    164783
      2322930         11380.000                    164783
      2323830          3292.000                    164783
      2323930          5412.000                    164783
      2325030          3662.000                    164783
      2325130          1410.000                    164783
      2512330          4683.000                    164783
      2750101            45.000                    164783
      2750201            36.780                    164783
      2751001         15960.000                    164783
      2751659          1359.000                    164783
      2762301            64.325                    164783
      2764001            78.000                    164783
      2771227           534.000                    164783
      2771559           255.000                    164783
      2803101           424.508                    164783
      2821501           880.000                    164783

In [43]:
medicaid_reimbursement_per_enrollee = np.log2(clean_temp["UNITS REIMBURSED"] / clean_temp["TOTAL MEDICAID ENROLLEES"])
medicaid_reimbursement_per_enrollee.head(20)

STATE  NDC    
AK     2143380    -9.609109
       2143480   -10.008280
       2322730    -6.109830
       2322830    -4.444321
       2322930    -3.855995
       2323830    -5.645459
       2323930    -4.928262
       2325030    -5.491792
       2325130    -6.868728
       2512330    -5.136991
       2750101   -11.838355
       2750201   -12.129358
       2751001    -3.368035
       2751659    -6.921878
       2762301   -11.322900
       2764001   -11.044806
       2771227    -8.269512
       2771559    -9.335854
       2803101    -8.600560
       2821501    -7.548848
dtype: float64

- To facilitate working with the final data, we are going to unstack `medicaid_reimbursement_per_enrollee` into a variable called  `medicaid_norm_ndc`

- Using `medicaid_reimbursement_per_enrollee`, generate a `DataFrame` where: 
  - index should be the two-letter state symbol 
  - the column names should be the NDC codes 

- The `DataFrame`  should be formatted as in the image below
  - Hint, simply unstack the data


<img src="media/unstacked.png" alt="drawing" style="width:900px;"/>


In [44]:
## WRITE YOUR CODE HERE 
medicaid_norm_ndc = medicaid_reimbursement_per_enrollee.unstack(level=-1)
medicaid_norm_ndc.head()

NDC    2143380      2143480      2322730      2322830      2322930      \
STATE                                                                    
AK       -9.609109   -10.008280    -6.109830    -4.444321    -3.855995   
AL       -9.940595    -9.805485    -5.336260    -4.243688    -3.645575   
AR      -12.985157   -12.657103    -4.720964    -2.886635    -2.617579   
AZ      -11.533870   -10.821194    -5.312806    -3.998085    -3.401510   
CA      -10.926863   -10.698372    -7.429936    -6.286911    -5.864352   

NDC    2323830      2323930      2325030      2325130      2512330      ...  \
STATE                                                                   ...   
AK       -5.645459    -4.928262    -5.491792    -6.868728    -5.136991  ...   
AL       -5.002524    -4.567483    -5.276300    -7.173792    -5.724359  ...   
AR       -3.834761    -3.496548    -3.946150    -5.755929    -4.833951  ...   
AZ       -4.857330    -4.103710    -4.737065    -5.898049    -6.346371  ...   
CA       -7.239140    -6.758794    -7.129391    -8.124464    -5.877665  ...   

NDC    43547039303  49884059002  63323020202  63323028002  63323028202  \
STATE                                                                    
AK             NaN          NaN          NaN          NaN          NaN   
AL             NaN          NaN          NaN          NaN          NaN   
AR             NaN          NaN          NaN          NaN          NaN   
AZ             NaN          NaN          NaN          NaN          NaN   
CA             NaN          NaN          NaN          NaN          NaN   

NDC    66758001604  68308002003  85433101     407141388    76329146901  
STATE                                                                   
AK             NaN          NaN          NaN          NaN          NaN  
AL             NaN          NaN          NaN          NaN          NaN  
AR             NaN          NaN          NaN          NaN          NaN  
AZ             NaN          NaN          NaN          NaN          NaN  
CA             NaN          NaN          NaN          NaN          NaN  

[5 rows x 21775 columns]

In [45]:
medicaid_norm_ndc.shape

(50, 21775)

#### Exploring the data (very briefly) 
- What is the drug with the highest log-normalized ratio in Hawaii?


In [46]:
## WRITE YOUR CODE HERE 
idx = medicaid_norm_ndc.loc['HI'].idxmax()
print('NDC of drug with the highest log-normalized ratio in Hawaii: ', idx)
drug_name = medicaid_data[medicaid_data['NDC'] == idx]['PRODUCT NAME'].unique()
print('Product name of drug with the highest log-normalized ratio in Hawaii: ', drug_name)

NDC of drug with the highest log-normalized ratio in Hawaii:  43386009019
Product name of drug with the highest log-normalized ratio in Hawaii:  ['GAVILYTE-G' 'GAVILYTE -' 'GaviLyte -']


* Investigate the NDC of the product with the highest log-normalized ratio in Hawaii 
  * What is it used for?

* Compare the value of `Units Reimbursed` that product between HI and other states, (take for instance MA, FL, OR and WA)
* Can you think for reasons why this product has the highest log-normalized ratio in Hawaii?

In [47]:
## WRITE YOUR CODE HERE 
#comp = medicaid_data[medicaid_data['STATE'].isin(['HI','MA','FL','OR','WA'])]
#comp[comp['NDC'] == idx][['STATE','UNITS REIMBURSED']]
medicaid_norm_ndc[[idx]].loc[['HI', 'MA', 'FL', 'OR', 'WA']]

NDC    43386009019
STATE             
HI       10.511181
MA       -8.395578
FL       -1.043195
OR        1.897731
WA       -7.736528

In [48]:
## WRITE YOUR ANSWER HERE 
agg_units = medicaid_data_w_enrollments_hierarch.groupby(level=[0,1])[['UNITS REIMBURSED']].sum()
agg_units.loc[['HI', 'MA', 'FL', 'OR', 'WA']].xs(idx, level=1).sort_values('UNITS REIMBURSED', ascending=False)

UNITS REIMBURSED
STATE                  
HI         4.969520e+08
OR         4.188000e+06
FL         1.848000e+06
WA         8.307267e+03
MA         5.433000e+03

### Answer 

The drug with NDC 43386009019 is 'Gavilyte-G'. While looking for it, I came to know that it is a prescription medicine given before colonoscopy. This may be needed if someone has colon cancer, irregular bowel or a specific lifestyle with no physical activity and bad dietary choices. On researching more, this site(https://www.gastroconsa.com/5-reasons-you-may-need-a-colonoscopy/) shows that eating a lot of pork or other processed meat by boiling or grilling can cause this. Since Hawaii has a lot of pork dishes, this might be one reason that it has the highest log-normalized ratio.

* Find and list all unique `NDC`s for which the difference between the largest and second large log-normalized ratio by state is at least 10.

* For instance:
  * The highest log-normalized ratio for `00591289749` (`AZACITIDIN`) is OK where it has a log-normalized ratio  of `-1.025642`.
  * The second highest log-normalized ratio for `00591289749` is in `GA` where is has a log-normalized ration of  `-12.623428`
  


In [49]:
## WRITE YOUR CODE HERE 
#highest_two = medicaid_norm_ndc.nlargest(2, columns = medicaid_norm_ndc.columns)
highest = medicaid_norm_ndc.apply(pd.Series.nlargest, args={1}, axis=0).min(axis=0)
second_highest = medicaid_norm_ndc.apply(pd.Series.nlargest, args={2}, axis=0).min(axis=0)
top2 = pd.concat([highest, second_highest], axis=1)
top2.columns = ['HIGHEST', 'SECOND HIGHEST']
top2['DIFFERENCE'] = top2['HIGHEST'] - top2['SECOND HIGHEST']
top2['DIFFERENCE'].loc[top2['DIFFERENCE']>=10].sort_values(ascending=False)

NDC
68180062210    16.514544
409767009      15.910902
407141233      12.441317
60793060002    12.023433
9023303        11.633238
591289749      11.597786
574082710      11.549606
121183605      10.600895
641047721      10.587633
338104902      10.196611
61703033218    10.169143
Name: DIFFERENCE, dtype: float64

- The Drug `AZACITIDINE` has a very high normalized UNITS REIMBURSED in OK compared to other states.
   - Normalized log value is -1.025642 (or a ratio 0.49119167009735065)
   - Second highest state has a log value of -12.623428 (0.000158478197834722)
- Oklahoma is not a high-incidence state for cancer
- Could the following explain what is happening in Oklahoma?

https://www.centerwatch.com/clinical-trials/listings/92093/acute-myeloid-leukemia-aml-study-asp2215-gilteritinib-by/?&geo_lat=35.4675602&geo_lng=-97.5164276&radius=10

### Answer
I am not sure since only a limited information is available but I'd guess that since this is clinical trial, to compare two different treatments, azacitidine and its combination with gilteritinib might be the reason for high log-normalized ratio.